clients/drcachesim/tests/invariant_checker_test.cpp

```diff
@@ -0,0 +1,331 @@
+
+#include "../tools/invariant_checker.h"
+#include "../common/memref.h"
+
+namespace {
+
+class checker_no_abort_t : public invariant_checker_t {
+public:
+    checker_no_abort_t(bool offline)
+        : invariant_checker_t(offline)
+    {
+    }
+    std::vector<std::string> errors;
+
+protected:
+    void
+    report_if_false(bool condition, const std::string &message) override
+    {
+        if (!condition)
+            errors.push_back(message);
+    }
+};
+
+memref_t
+gen_instr_type(trace_type_t type, memref_tid_t tid, addr_t pc)
+{
+    memref_t memref = {};
+    memref.instr.type = type;
+    memref.instr.tid = tid;
+    memref.instr.addr = pc;
+    memref.instr.size = 1;
+    return memref;
+}
+
+memref_t
+gen_instr(memref_tid_t tid, addr_t pc)
+{
+    return gen_instr_type(TRACE_TYPE_INSTR, tid, pc);
+}
+
+memref_t
+gen_branch(memref_tid_t tid, addr_t pc)
+{
+    return gen_instr_type(TRACE_TYPE_INSTR_CONDITIONAL_JUMP, tid, pc);
+}
+
+memref_t
+gen_marker(memref_tid_t tid, trace_marker_type_t type, uintptr_t val)
+{
+    memref_t memref = {};
+    memref.marker.type = TRACE_TYPE_MARKER;
+    memref.marker.tid = tid;
+    memref.marker.marker_type = type;
+    memref.marker.marker_value = val;
+    return memref;
+}
+
+memref_t
+gen_exit(memref_tid_t tid)
+{
+    memref_t memref = {};
+    memref.instr.type = TRACE_TYPE_THREAD_EXIT;
+    memref.instr.tid = tid;
+    return memref;
+}
+
+bool
+check_branch_target_after_branch()
+{
+    // Positive simple test.
+    {
+        checker_no_abort_t checker(true /*offline*/);
+        std::vector<memref_t> memrefs = {
+            gen_instr(1, 1), gen_branch(1, 2),
+            gen_instr(1, 3), gen_marker(2, TRACE_MARKER_TYPE_TIMESTAMP, 0),
+            gen_instr(2, 1),
+        };
+        for (const auto &memref : memrefs) {
+            checker.process_memref(memref);
+        }
+        for (const auto &error : checker.errors) {
+            std::cerr << "Unexpected error: " << error << "\n";
+        }
+        if (!checker.errors.empty())
+            return false;
+    }
+    // Negative simple test.
+    {
+        checker_no_abort_t checker(true /*offline*/);
+        std::vector<memref_t> memrefs = {
+            gen_instr(1, 1),
+            gen_branch(1, 2),
+            gen_marker(2, TRACE_MARKER_TYPE_TIMESTAMP, 0),
+            gen_instr(2, 1),
+        };
+        for (const auto &memref : memrefs) {
+            checker.process_memref(memref);
+        }
+        if (checker.errors.size() != 1 ||
+            checker.errors[0] != "Branch target not immediately after branch") {
+            std::cerr << "Failed to catch bad branch target poisition\n";
+            return false;
+        }
+    }
+    // Invariant relaxed for thread exit or signal.
+    {
+        checker_no_abort_t checker(true /*offline*/);
+        std::vector<memref_t> memrefs = {
+            gen_marker(3, TRACE_MARKER_TYPE_CACHE_LINE_SIZE, 64),
+            gen_branch(3, 2),
+            gen_exit(3),
```

Comment:
Curious when this happens: abrupt exit after a branch. Shouldn't there be other instructions leading up to an exit system call?

Reply:
If we traced the entire application lifetime, yes, but for detach the exit point can be anywhere our signal arrived: and since we delay asynch signals to block end it is almost always after a branch in practice.